sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala

```diff
@@ -258,6 +258,26 @@ object FileSourceStrategy extends Strategy with PredicateHelper with Logging {
       val outputAttributes = readDataColumns ++ generatedMetadataColumns ++
         partitionColumns ++ constantMetadataColumns
 
+      // Rebind metadata attribute references in filters after the metadata attribute struct has
+      // been flattened. Only data filters can contain metadata references. The rebinding will
+      // categorize the references as either [[FileSourceConstantMetadataAttribute]] or
+      // [[FileSourceGeneratedMetadataAttribute]].
+      def rebindFileSourceMetadataAttributesInFilters(
+          filters: Seq[Expression]): Seq[Expression] =
+        filters.map { filter =>
+          filter.transform {
+            case fieldGetter @ GetStructField(FileSourceMetadataAttribute(_), _, _) =>
+              val fieldName = fieldGetter.extractFieldName
+              // The row_index column has been renamed.
+              val columnName = if (fieldName == FileFormat.ROW_INDEX) {
+                FileFormat.ROW_INDEX_TEMPORARY_COLUMN_NAME
+              } else {
+                fieldName
+              }
+              metadataColumns.find(_.name == columnName).get
```

Comment:
shall we consider case sensitivity?

Reply:
The flattening does not change the name, so I guess not doing that is a bit more safe on making this a 1 to 1 comparison? On the other hand, duplicate names because of different casing are not allowed AFAIK